### DISCUSSION TITLE
I can't enable CORS

### DISCUSSION THREAD

**Discussion by fedecastelli** [fastapi/fastapi#7880] [Questions] [answered]
I set up a Dockerimage that uses fastapi version 0.45.0  and uvicorn. I need to enable CORS for different domains like `id-localtest.mydomain.com`, `id-alpha.mydomain.com` and `id.mydomain.com`. In my `main.py` file I wrote these lines of code

```python
from fastapi import FastAPI, HTTPException

app = FastAPI()

origins = ['https://id-localtest.mydomain.com', 'https://id-alpha.mydomain.com', 'https://id.mydomain.com']

app.add_middleware(
    CORSMiddleware,
    allow_origins=origins,
    allow_credentials=True,
    allow_methods=["DELETE,GET,POST,PUT"],
    allow_headers=["*"])

@app.post("/{stage}/activation")
async def activation_post(stage: str, request: Request):
    # my custom code
    return 200
```
I'm always getting an HTTP status code 400 when a generic browser makes an OPTIONS request. 
If I try to use Postman to make an OPTIONS request, I get the HTTP status code `405` and the json received is:

```json
{
    "detail": "Method Not Allowed"
}
```
What am I doing wrong? I'm following this tutorial https://fastapi.tiangolo.com/tutorial/cors/

**Comment by nsidnev** [comment, marked as answer]
[`starlette.middleware.cors.CORSMiddleware`](https://github.com/encode/starlette/blob/7bcb07786a538596441c563af7ea935c810c5336/starlette/middleware/cors.py#L17) accepts a sequence of strings in `allow_methods` parameter. Try replacing your line with:
```python 
    allow_methods=["DELETE", "GET", "POST", "PUT"],
```

Also, your browser does not send a simple `OPTIONS` request, but a [preflight-request](https://developer.mozilla.org/en-US/docs/Glossary/Preflight_request). You can send it manually by specifying special headers for your request. Here is an example for your application using `httpie`:
```bash
$ http --verbose OPTIONS :8000/stage/activation Access-Control-Request-Method:POST Origin:https://id-localtest.mydomain.com

OPTIONS /stage/activation HTTP/1.1
Accept: */*
Accept-Encoding: gzip, deflate
Access-Control-Request-Method: POST
Connection: keep-alive
Content-Length: 0
Host: localhost:8000
Origin: https://id-localtest.mydomain.com
User-Agent: HTTPie/1.0.3



HTTP/1.1 200 OK
access-control-allow-credentials: true
access-control-allow-methods: DELETE, GET, POST, PUT
access-control-allow-origin: https://id-localtest.mydomain.com
access-control-max-age: 600
content-length: 2
content-type: text/plain; charset=utf-8
date: Tue, 07 Jan 2020 15:19:54 GMT
server: uvicorn
vary: Origin

OK
```

**Comment by fedecastelli** [comment]
Oh, my fault! I didn't notice it! I just copied and pasted the methods accepted by Flask thinking I could specify them in the same way. Thank you!

**Comment by tiangolo** [comment]
Thanks for the help here @nsidnev ! :clap: :bow: 

And thanks @fedecastelli for reporting back and closing the issue :+1:

**Comment by ahmedbr** [comment]
Hi all,
Although I'm using the same code as above, I'm facing a similar problem as following:
`Status Code: 422 Unprocessable Entity
Referrer Policy: strict-origin-when-cross-origin
`
Any ideas to solve the problem?

**Comment by mauroalejandrojm** [comment]
> Hi all,
> Although I'm using the same code as above, I'm facing a similar problem as following:
> `Status Code: 422 Unprocessable Entity Referrer Policy: strict-origin-when-cross-origin `
> Any ideas to solve the problem?

Hello, 422 Unprocessable Entity is not a CORS error, can you check your the body request for any "," or wrong fields that are being validated.

**Comment by ahmedbr** [comment]
> > Hi all,
> > Although I'm using the same code as above, I'm facing a similar problem as following:
> > `Status Code: 422 Unprocessable Entity Referrer Policy: strict-origin-when-cross-origin `
> > Any ideas to solve the problem?
> 
> Hello, 422 Unprocessable Entity is not a CORS error, can you check your the body request for any "," or wrong fields that are being validated.

The problem has been solved after creating a subdomain for my web service.

**Comment by hanzohan** [comment]
please note that you can also add 

- main domain
- sub domain
- different port

in origins sections, for example: 

```
origins = [
    "https://aaa.com", # main domain, targeting HTTPS port [433] by default because we don't specify the port.
    "https://abc.aaa.com", # subdomain, targeting HTTPS port [433] by default because we don't specify the port.
    "https://aaa.com:4000" # main domain, targeting 4000 port
]
```
you can also allow all standard methods by set 
`allow_methods=["*"]`

for example: 
```
app.add_middleware(
    CORSMiddleware,
    allow_origins=origins,
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)
```

**Comment by jeremyalv** [comment]
If you've added middleware correctly with all the options, origins etc. but it still doesn't work, try checking if you are putting the add_middleware() **above** the include_route() if you use different files for routing - middleware order matters. Took me 2 days to figure out that my add_middleware() was placed below. Hope this saves someone's day.
